middle_end/flambda2/to_cmm/to_cmm_env.ml

```diff
@@ -208,150 +323,407 @@ let get_exn_extra_args env k =
+  | Split _ -> res, Already_split
+  | Splittable_prim { dbg; prim; args } ->
+    let new_bindings, new_cmm_args =
+      new_bindings_for_splitting binding.order args
+    in
+    let new_cmm_expr, res = rebuild_prim ~dbg ~env ~res prim new_cmm_args in
+    let prim_effects =
+      Flambda_primitive.Without_args.effects_and_coeffects prim
+    in
+    (match To_cmm_effects.classify_by_effects_and_coeffects prim_effects with
+    | Pure | Generative_immutable -> ()
+    | Effect | Coeffect_only ->
+      Misc.fatal_errorf
+        "Once split, a 'must_inline_once' binding cannot have effects or \
+         coeffects, since it can be moved around to be inlined.");
+    let split_binding =
+      { order = binding.order;
+        effs = prim_effects;
+        inline = binding.inline;
+        bound_expr = Split { cmm_expr = new_cmm_expr };
+        cmm_var = binding.cmm_var
+      }
+    in
+    (*
+     * Format.eprintf "*** new bindings:@\n@[<v>%a@]@."
+     *   (Format.pp_print_list
+     *     ~pp_sep:(fun ppf () -> Format.fprintf ppf "@,")
+     *      print_any_binding)
+     *  new_bindings;
+     * Format.eprintf "*** split_binding:@\n%a@\n@." print_binding split_binding;
+     *)
+    res, Split { new_bindings; split_binding }
+
+let remove_binding env var =
+  { env with bindings = Variable.Map.remove var env.bindings }
+
+let will_inline_simple env res { effs; bound_expr = Simple { cmm_expr }; _ } =
+  cmm_expr, env, res, effs
+
+let will_inline_complex env res { effs; bound_expr; _ } =
+  match bound_expr with
+  | Split { cmm_expr } -> cmm_expr, env, res, effs
+  | Splittable_prim { dbg; prim; args } ->
+    let cmm_expr, res = rebuild_prim ~dbg ~env ~res prim (List.map fst args) in
+    cmm_expr, env, res, effs
+
+let will_not_inline_simple env res { cmm_var; bound_expr = Simple _; _ } =
+  ( C.var (Backend_var.With_provenance.var cmm_var),
+    env,
+    res,
+    Ece.pure_can_be_duplicated )
+
+let split_and_inline env res var binding =
+  let res, split_result = split_complex_binding ~env ~res binding in
+  match split_result with
+  | Already_split -> will_inline_complex env res binding
+  | Split { new_bindings; split_binding } ->
+    let env =
+      (* for duplicated bindings, we need to replace the original splittable
+         binding with the new split binding in the bindings map of the env *)
+      match split_binding.inline with
+      | Must_inline_once -> env
+      | Must_inline_and_duplicate ->
+        { env with
+          bindings = Variable.Map.add var (Binding split_binding) env.bindings
+        }
+    in
+    let env =
+      List.fold_left
+        (fun env new_binding ->
+          let flambda_var = Variable.create "to_cmm_tmp" in
+          { env with
+            bindings = Variable.Map.add flambda_var new_binding env.bindings
+          })
+        env new_bindings
+    in
+    will_inline_complex env res split_binding
+
+let pop_from_top_stage ?consider_inlining_effectful_expressions env var =
```

Comment:
I wish the name was more explicit. We're not unconditionally popping, but only if the given variable is at the top of the stages.

Reply:
renamed to `pop_if_in_top_stage` which is more explicit